### PR TITLE
Telecommand to fetch the current active agenda

### PR DESCRIPTION
**Pull request by christopherkinyua**
This PR is for creating a telecommand that prints out the active agenda in a JSONL format

I achieved this by:
- Looped through the entire agendas and counted the active agendas to find the total active and printed the number out
- Looped through the entire agendas again, parsed the data into a string, and used the `LOG_message` to display the results
- The outputted data is: `slot_num`,  `tcmd_channel`, `timestamp_sent` and `timestamp_to_execute`. I tried using the tcmd_idx but it was not printing out a value hence I decided to use the slot_num instead.

Let me know if there is a better way to do this and I would appreciate your review.

### CODE REVIEW

**Comment by parker-research** [comment]
It looks like there's some merge conflicts. Can you please pull the latest version of `main`, and then rebase your branch to main first? Thanks!

**Comment by christopherkinyua** [comment]
Okay, Im trying to resolve the conflicts as we speak

File: firmware/Core/Inc/telecommands/agenda_telecommands.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#ifndef __INCLUDE_GUARD__AGENDA_TELECOMMAND_H
+#define __INCLUDE_GUARD__AGENDA_TELECOMMAND_H
+
+#include "telecommands/telecommand_types.h"
```
Comment (parker-research) [inline firmware/Core/Inc/telecommands/agenda_telecommands.h:4]: Should this go in the `agenda_telecommand_defs.h` file?

File: firmware/Core/Src/telecommands/telecommand_definitions.c [hunk 2]
```diff
@@ -332,6 +333,17 @@ const TCMD_TelecommandDefinition_t TCMD_telecommand_definitions[] = {
         .readiness_level = TCMD_READINESS_LEVEL_FOR_OPERATION,
     },
 
+    // ****************** SECTION: agenda_telecommand_defs ******************
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/telecommand_definitions.c:336]: Doesn't this section already exist in this file?

File: firmware/Core/Src/telecommands/telecommand_executor.c [hunk 3]
```diff
@@ -18,6 +21,18 @@ TCMD_parsed_tcmd_to_execute_t TCMD_agenda[TCMD_AGENDA_SIZE];
 ///         (i.e., filled with a not-yet-executed command).
 uint8_t TCMD_agenda_is_valid[TCMD_AGENDA_SIZE] = {0};
 
+/// @brief Converts a TCMD_TelecommandChannel_enum_t to a string representation.
+/// @param channel Input TCMD_TelecommandChannel_enum_t
+/// @return A pointer to a C-string representing the TCMD_TelecommandChannel_enum_t.
+const char* getTelecommandChannelName(TCMD_TelecommandChannel_enum_t channel) {
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/telecommand_executor.c:27]: Bad name format. Please use snake case, with proper prefix.

Great function to create though!

File: firmware/Core/Src/telecommands/telecommand_executor.c [hunk 4]
```diff
@@ -119,12 +134,9 @@ uint8_t TCMD_execute_parsed_telecommand_now(const uint16_t tcmd_idx, const char
     }
     TCMD_TelecommandDefinition_t tcmd_def = TCMD_telecommand_definitions[tcmd_idx];
 
-    DEBUG_uart_print_str("=========================");
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/telecommand_executor.c:122]: Please revert these changes in this PR. They're unnecessary (and maybe a regression).

File: firmware/Core/Src/telecommands/telecommand_executor.c [hunk 5]
```diff
@@ -240,4 +251,65 @@ uint8_t TCMD_agenda_delete_by_tssent(uint64_t tssent) {
...
+    // Count the number of active agendas
+    for (uint16_t slot_num = 0; slot_num < TCMD_AGENDA_SIZE; slot_num++) {
+        if (TCMD_agenda_is_valid[slot_num]) {
+            active_agendas++;
+        }
+    }
+
+    // if no active agendas, return 1
+    if(active_agendas == 0){
+        LOG_message( LOG_SYSTEM_TELECOMMAND, LOG_SEVERITY_NORMAL, LOG_SINK_ALL,
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/telecommand_executor.c:272]: Please fix indentation.

File: firmware/Core/Src/telecommands/telecommand_executor.c [hunk 6]
```diff
@@ -240,4 +251,65 @@ uint8_t TCMD_agenda_delete_by_tssent(uint64_t tssent) {
...
+                slot_num,
+                getTelecommandChannelName(TCMD_agenda[slot_num].tcmd_channel),
+                tssent_str,
+                tsexec_str
+            );
+        }
+            LOG_message( LOG_SYSTEM_TELECOMMAND, LOG_SEVERITY_NORMAL, LOG_SINK_ALL, message_buffer);
+            logged_agendas++;
+
+            // Break the loop once all active agendas have been logged
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/telecommand_executor.c:307]: Please note that this is an "early-exit optimization" in a comment

File: firmware/Core/Src/telecommands/agenda_telecommands.c [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+#include <string.h>
+#include <stdio.h>
+
+/// @brief Telecommand: Fetch all active agendas
+/// @param args_str No arguments needed
+/// @param tcmd_channel The channel on which the telecommand was received, and on which the response should be sent
+/// @param response_output_buf The buffer to write the response to
+/// @param response_output_buf_len The maximum length of the response_output_buf (its size)
+/// @return 0 on success, 1 if there are no active agendas.
+uint8_t TCMDEXEC_agenda_fetch(const char *args_str, TCMD_TelecommandChannel_enum_t tcmd_channel,
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/agenda_telecommands.c:16]: Please rename to `TCMDEXEC_agenda_fetch_jsonl` to specify the result format.

File: firmware/Core/Src/telecommands/agenda_telecommands.c [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#include "telecommands/telecommand_definitions.h"
+#include "telecommands/telecommand_args_helpers.h"
+#include "telecommands/telecommand_executor.h"
+#include "debug_tools/debug_uart.h"
+#include "telecommands/agenda_telecommands.h"
+
+#include <string.h>
+#include <stdio.h>
+
+/// @brief Telecommand: Fetch all active agendas
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/agenda_telecommands.c:10]: Change to "Fetch all pending agenda items, and log them each as JSONL"

**Review by parker-research — CHANGES_REQUESTED** [review]
* Please fix indentation. Indent using spaces.
* Looks like you're adding stuff to a new file that should go in an existing file (comments within).

Awesome work though!

**Comment by christopherkinyua** [comment]
Ohh shoot. I see that. It must have come to when I was rebasing the branch last night. I will migrate the functions and code into `agenda_telecommand_defs.c/h`

File: firmware/Core/Src/telecommands/agenda_telecommands.c (hunk 8 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/agenda_telecommands.c:10]: Done

File: firmware/Core/Src/telecommands/telecommand_definitions.c (hunk 2 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/telecommand_definitions.c:336]: Yeah it does, I fixed it and renamed it

File: firmware/Core/Inc/telecommands/agenda_telecommands.h (hunk 1 above)
Comment (christopherkinyua) [reply firmware/Core/Inc/telecommands/agenda_telecommands.h:4]: Migrated code to `agenda_telecommand_defs.h/c` files

File: firmware/Core/Src/telecommands/telecommand_executor.c (hunk 4 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/telecommand_executor.c:122]: Sure thing. I just copied what is originally in main into this

File: firmware/Core/Src/telecommands/telecommand_executor.c (hunk 3 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/telecommand_executor.c:27]: Changed it to `telecommand_channel_enum_to_str` instead

File: firmware/Core/Src/telecommands/telecommand_executor.c (hunk 5 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/telecommand_executor.c:272]: I made changes to the indentation

File: firmware/Core/Src/telecommands/agenda_telecommands.c (hunk 7 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/agenda_telecommands.c:16]: Done

File: firmware/Core/Src/telecommands/telecommand_executor.c (hunk 6 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/telecommand_executor.c:307]: Sure thing. I added the early-exit optimization into the comment

**Comment by christopherkinyua** [comment]
I also noticed in line 185-189 when we are converting the uint64_t to a string, for the `tssent_str` we are passing the `.timestamp_to_execute` instead of the `.timestamp_sent`. This is in the `TCMD_execute_telecommand_in_agenda` function in `telecommand_executor.c` file.

I am not sure if this was done intentionally or it was a mistake.

File: firmware/Core/Src/telecommands/agenda_telecommand_defs.c [hunk 9]
```diff
@@ -66,3 +66,16 @@ uint8_t TCMDEXEC_agenda_delete_by_tssent(const char *args_str, TCMD_TelecommandC
...
+/// @brief Telecommand: Fetch all pending agenda items, and log them each as JSONL
+/// @param args_str No arguments needed
+/// @param tcmd_channel The channel on which the telecommand was received, and on which the response should be sent
+/// @param response_output_buf The buffer to write the response to
+/// @param response_output_buf_len The maximum length of the response_output_buf (its size)
+/// @return 0 on success, 1 if there are no active agendas.
+uint8_t TCMDEXEC_agenda_fetch_jsonl(const char *args_str, TCMD_TelecommandChannel_enum_t tcmd_channel,
+                        char *response_output_buf, uint16_t response_output_buf_len) {
+     
+    uint8_t result = TCMD_agenda_fetch();
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/agenda_telecommand_defs.c:79]: Best to make this a `const`

File: firmware/Core/Src/telecommands/telecommand_executor.c [hunk 10]
```diff
@@ -240,4 +253,66 @@ uint8_t TCMD_agenda_delete_by_tssent(uint64_t tssent) {
...
+uint8_t TCMD_agenda_fetch(){
+    uint16_t active_agendas = 0;
+    uint16_t logged_agendas = 0;
+    char message_buffer[512];
+    
+    // Count the number of active agendas
+    for (uint16_t slot_num = 0; slot_num < TCMD_AGENDA_SIZE; slot_num++) {
+        if (TCMD_agenda_is_valid[slot_num]) {
+            active_agendas++;
+            }
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/telecommand_executor.c:269]: Bad indent.

File: firmware/Core/Src/telecommands/telecommand_executor.c [hunk 11]
```diff
@@ -240,4 +253,66 @@ uint8_t TCMD_agenda_delete_by_tssent(uint64_t tssent) {
...
+                slot_num,
+                telecommand_channel_enum_to_str(TCMD_agenda[slot_num].tcmd_channel),
+                tssent_str,
+                tsexec_str
+            );
+
+            LOG_message( LOG_SYSTEM_TELECOMMAND, LOG_SEVERITY_NORMAL, LOG_SINK_ALL, message_buffer);
+        }
+
+            logged_agendas++;
```
Comment (parker-research) [inline firmware/Core/Src/telecommands/telecommand_executor.c:308]: Bad indentation

**Review by parker-research — CHANGES_REQUESTED** [review]
This looks super great now! Just a few style tweaks

I think you identified a mistake with your latest comment. Can you please add it as a fix in its own commit in this PR? Thanks :)

**Comment by christopherkinyua** [comment]
Currently working on addressing the changes. I will message when I am done

File: firmware/Core/Src/telecommands/agenda_telecommand_defs.c (hunk 9 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/agenda_telecommand_defs.c:79]: Done

File: firmware/Core/Src/telecommands/telecommand_executor.c (hunk 11 above)
Comment (christopherkinyua) [reply firmware/Core/Src/telecommands/telecommand_executor.c:308]: Done

**Comment by christopherkinyua** [comment]
PR is ready for your review

**Review by parker-research — APPROVED** [review]
Very awesome overall! 

I just tacked on a commit which fixed the indentation issue, and changed the usage of `LOG_message` to use its format string capability.

If you're happy with it, please Squash and Merge!

Again, great catch on that `tssent` mixup!

**Comment by christopherkinyua** [comment]
Ohh nice, thats a cool feature of the `LOG_message`. What was the indentation problem? My apologies for still not seeing it.

Ill squash and merge

**Comment by parker-research** [comment]
See the indentation changes made here: https://github.com/CalgaryToSpace/CTS-SAT-1-OBC-Firmware/pull/161/commits/a48b026aef033d518852b1eaf0fa675793b2be6e, if you're curious

**Comment by christopherkinyua** [comment]
Ahhh I see, the indentation is with `LOG_message`. Ill try and do it in the future.
